rmw_zenoh_cpp/src/detail/event.hpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DETAIL__EVENT_HPP_
+#define DETAIL__EVENT_HPP_
+
+#include <condition_variable>
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+#include "rmw/event.h"
+#include "rmw/event_callback_type.h"
+
+
+///=============================================================================
+// A struct that represents an event status in rmw_zenoh.
+enum rmw_zenoh_event_type_t
+{
+  // sentinel value
+  ZENOH_EVENT_INVALID,
+
+  // subscription events
+  ZENOH_EVENT_REQUESTED_QOS_INCOMPATIBLE,
+  ZENOH_EVENT_MESSAGE_LOST,
+  ZENOH_EVENT_SUBSCRIPTION_INCOMPATIBLE_TYPE,
+  ZENOH_EVENT_SUBSCRIPTION_MATCHED,
+
+  // publisher events
+  // RMW_EVENT_LIVELINESS_LOST,
+  // RMW_EVENT_OFFERED_DEADLINE_MISSED,
+  ZENOH_EVENT_OFFERED_QOS_INCOMPATIBLE,
+  ZENOH_EVENT_PUBLISHER_INCOMPATIBLE_TYPE,
+  ZENOH_EVENT_PUBLICATION_MATCHED,
+};
+
+/// Helper value to indicate the maximum index of events supported.
+#define ZENOH_EVENT_ID_MAX rmw_zenoh_event_type_t::ZENOH_EVENT_PUBLICATION_MATCHED
```

Comment:
This is pretty minor, but our typical style here is to add a "SENTINEL" value at the end of the enum, and check against that instead.  So this would change to:

```suggestion
  ZENOH_EVENT_INVALID
};
```

And then the checks in `event.cpp` would change to `>= ZENOH_EVENT_INVALID`.

Reply:
I see. If we adopt this approach, the array lengths of these [members](https://github.com/ros2/rmw_zenoh/blob/e51e475760347bbd6dcb4d79b7a349e9264a0af4/rmw_zenoh_cpp/src/detail/event.hpp#L159-L170) would now be `ZENOH_EVENT_INVALID`, eg `std::condition_variable * event_conditions_[ZENOH_EVENT_INVALID]{nullptr}` which presents a readability problem imo. If you think it's valid, I'm happy to make the change as suggested.